keysync_initiator.go

```diff
@@ -0,0 +1,174 @@
+package nitriding
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hf/nitrite"
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+// RequestKeys asks a remote enclave to share its key material with us, which
+// is then written to the provided variable.
+//
+// This is only necessary if you intend to scale enclaves using Kubernetes.  If
+// you will only ever run a single enclave, ignore this function.
+func RequestKeys(addr string, keyMaterial any) error {
+	errStr := "failed to request key material"
+
+	// First, request a nonce from the remote enclave.
+	theirNonce, err := requestNonce(addr)
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Now, create our own nonce.
+	ourNonce, err := newNonce()
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Next, create a key that the remote enclave is going to use to encrypt
+	// its key material.
+	sbKey, err := newSbKey()
```

Comment:
The paper mentions public key cryptography for this exchange, but secretbox says it uses 'secret-key cryptography'. Is this weaker than the paper's claim or am I misunderstanding something?

Using a key pair here seems better, since an attacker who has broken TLS can poison the new enclave, but not decrypt key material from a running enclave.

Reply:
Might as well. I'm going to address this in a later commit.

Reply:
https://pkg.go.dev/filippo.io/age maybe?

Reply:
or https://pkg.go.dev/golang.org/x/crypto/nacl/box

Reply:
I have an almost-finished commit based on box. Will push later today.

Reply:
Fixed in 45ab730. The following has changed:

- Use the box instead of the secretbox API.
- Get rid of the secretbox-specific nonce.
- Replace the single secret key with a public and private key pair.

Subsequent commits clean up the code a bit and add documentation and tests.